### ISSUE TITLE
Gradient Descent & Backpropagation

### ISSUE THREAD

**Issue by pillarliang** [pillarliang/blog#9]
# Gradient Descent
## 梯度

在数学中：梯度是一个包含多元函数对其所有变量的偏导数的向量

<img width="345" alt="12" src="https://github.com/pillarliang/blog/assets/28833267/40d54941-f5ed-4bf0-922c-7d2bb41873d2">

梯度表示的是各点处的函数值变化最大的方向。

在**梯度法**中，函数的取值从当前位置沿着梯度方向前进一定距离，然后在新的地方重新求梯度，再沿着新梯度方向前进， 如此反复，不断地沿梯度方向前进。迭代更新直至梯度 g 为 0，但是 g 到 0非常难，实际操作中不会算到 g 为 0，都是自己决定什么时候停下来不更新了。像这样，通过不断地沿梯度方向前进， 逐渐减小函数值的过程就是**梯度法（gradient method）**

只需理解思想，实际应用中直接使用框架就好。

## 思路

我们是对一堆参数 $\theta = \{ w_1,w_2,...b_1,b_2\}$ 做梯度下降。先选定一个初始参数 $\theta^0$。计算 network 里面每一个参数对 $L(\theta^0)$ 的偏微分得到 $∇L(\theta^0)$。继而继续算 $\theta^1 = \theta^0 - \eta∇L(\theta^0)$;

$\theta^2 = \theta^1 - \eta∇L(\theta^1)$ 一直update 下去。但是要如何有效地去把这一个很大（百万维）的 vector有效地把它计算出来，就用到 `Backpropagation`。

## **移动的方向和移动的大小**

这里所说的梯度是指损失函数关于权重参数的梯度。

$$
w^1 = w^0 - \eta {\partial L \over \partial w}|_{w=w^0}
$$

<img width="465" alt="2" src="https://github.com/pillarliang/blog/assets/28833267/a9d52074-f64b-45c3-bf32-9cf1190c8810">

<img width="465" alt="3" src="https://github.com/pillarliang/blog/assets/28833267/7f7d5c12-b561-46ac-8290-72614405c09f">

### **方向**：

${\partial L \over \partial w}|_{w=w^0}$

上面的内容中移动方向只是一个概述，不是准确的表达。

- 梯度为正（positive），向相反的方向移动，即 increase 自变量
- 梯度为负（negative），向相同的方向移动，即 decrease 自变量

**计算**：全部变量（feature）的偏导数汇总而成的向量称为梯度（gradient）。

### **大小**

通过`学习率 $\eta$` 控制。学习率是hyperparameter（超参数），需要自己设定的参数，来调整跨越步伐的大小。

> 像学习率这样的参数称为**超参数**。这是一种和神经网络的参数（权重 和偏置）性质不同的参数。相对于神经网络的权重参数是通过训练 数据和学习算法自动获得的，学习率这样的超参数则是人工设定的。 一般来说，超参数需要尝试多个值，以便找到一种可以使学习顺利 进行的设定。调整 **超参数** 的过程叫调参。
> 

## 分类

梯度下降算法分为批量梯度下降(Batch Gradient Descent)、随机梯度下降(Stochastic Gradient Descent)和小批量梯度下降(Mini-Batch Gradient Descent)三种算法。其中：

- **批量梯度下降**使用所有样本进行更新
- **小批量梯度下降**则使用一定批量的样本进行更新。
- **Stochastic Gradient Descent update using one sample**

**存在问题：当到达 local minima 时停止更新**

<img width="465" alt="5" src="https://github.com/pillarliang/blog/assets/28833267/1c71a5a4-f26b-4103-9405-bb1fe74d6b48">

**optimization 的实际操作：**

实际操作不会去算整个 data 的 loss 然后才求梯度，而是根据 `batch size` 将data分成很多 `batch`，每次更新都只用一批（`batch`）的 data 来算梯度！下次更新的时候又换一批资料算梯度。当看完了整个资料，也就是所有 `batch` 都使用到时，就叫一个 `epoch`。

batch size 也是一个 hyper parameter。

> 目前学到的 hyper parameter（自己设定的参数）：learning rate、sigmoid数量、batch size。
> 

## Step

1. Take the Gradient of the Loss Function.
2. Pick random values for the parameters, and then Plug the parameter values into the derivatives (ahem, the Gradient).
3. Calculate the Step Sizes: Step Size = Slope × Learning Rate.
4. Calculate the New Parameters: New Parameter = Old Parameter - Step Size

# Gradient Descent
`Backpropagation` 是一个是有效计算梯度的方法。

因为更新参数时，需要计算 `Loss` 关于该参数的偏微分，神经网络的 layer，要计算当前layer 的参数偏微分，需要走完整个神经网络层得到最后的输出结果来计算。这个计算的方法叫`Backpropagation`

$$
w^1 = w^0 - \eta {\partial L \over \partial w}|_{w=w^0}
$$

反向传播算法是神经网络中**训练模型**的核心算法之一。在神经网络中，最终的输出结果和目标结果之间存在误差，需要通过优化模型参数（权值和偏差）来最小化误差。反向传播算法是实现这个过程的关键。

在**反向传播算法**中，首先通过**前向传播**的方式计算模型的输出结果，并与目标结果进行比较，得到**误差**。然后，根据误差反向计算每一层的梯度，更新模型的参数。

<img width="502" alt="6" src="https://github.com/pillarliang/blog/assets/28833267/30d3c204-af11-46ab-9537-2b5dca7a9203">
<img width="502" alt="7" src="https://github.com/pillarliang/blog/assets/28833267/9d45db60-0a10-4008-9707-c4269d192637">
